mkosi/qemu.py

```diff
@@ -1503,25 +1519,31 @@ def add_virtiofs_mount(
 
         if QemuDeviceNode.vhost_vsock in qemu_device_fds:
             addr, notify = stack.enter_context(vsock_notify_handler())
-            credentials["vmm.notify_socket"] = addr
+            (credentials / "vmm.notify_socket").write_text(addr)
 
         if config.forward_journal:
-            credentials["journal.forward_to_socket"] = stack.enter_context(
-                start_journal_remote_vsock(config)
+            (credentials / "journal.forward_to_socket").write_text(
+                stack.enter_context(start_journal_remote_vsock(config))
             )
 
-        for k, v in credentials.items():
-            payload = base64.b64encode(v.encode()).decode()
+        smbiosdir = Path(stack.enter_context(tempfile.TemporaryDirectory(prefix="mkosi-smbios-")))
+
+        for p in credentials.iterdir():
+            payload = base64.b64encode(p.read_text().encode()).decode()
+
             if config.architecture.supports_smbios(firmware):
-                cmdline += ["-smbios", f"type=11,value=io.systemd.credential.binary:{k}={payload}"]
+                with (smbiosdir / p.name).open("w") as f:
+                    f.write(f"io.systemd.credential.binary:{p.name}=")
+                    f.write(payload)
+
+                cmdline += ["-smbios", f"type=11,path={smbiosdir / p.name}"]
             # qemu's fw_cfg device only supports keys up to 55 characters long.
-            elif config.architecture.supports_fw_cfg() and len(k) <= 55 - len("opt/io.systemd.credentials/"):
-                f = stack.enter_context(tempfile.NamedTemporaryFile(prefix="mkosi-fw-cfg-", mode="w"))
-                f.write(v)
-                f.flush()
-                cmdline += ["-fw_cfg", f"name=opt/io.systemd.credentials/{k},file={f.name}"]
+            elif config.architecture.supports_fw_cfg() and len(p.name) <= 55 - len(
+                "opt/io.systemd.credentials/"
+            ):
+                cmdline += ["-fw_cfg", f"name=opt/io.systemd.credentials/{p.name},file={p}"]
             elif kernel:
-                kcl += [f"systemd.set_credential_binary={k}:{payload}"]
+                kcl += [f"systemd.set_credential_binary={p.name}:{payload}"]
```

Comment:
This is a footgun:
```
>>> bar=b"bar"
>>> f"foo={bar}"
"foo=b'bar'"
```
i.e. when putting a byte into an f-string, this will get stringified in an unexpected way
```suggestion
                kcl += [f"systemd.set_credential_binary={p.name}:{payload.decode()}"]
```